x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/queries/SemanticQueryRewriteInterceptor.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.inference.queries;
+
+import org.elasticsearch.action.ResolvedIndices;
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.metadata.InferenceFieldMetadata;
+import org.elasticsearch.index.mapper.IndexFieldMapper;
+import org.elasticsearch.index.query.BoolQueryBuilder;
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.index.query.QueryRewriteContext;
+import org.elasticsearch.index.query.TermsQueryBuilder;
+import org.elasticsearch.plugins.internal.rewriter.QueryRewriteInterceptor;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Intercepts and adapts a query to be rewritten to work seamlessly on a semantic_text field.
+ */
+public abstract class SemanticQueryRewriteInterceptor implements QueryRewriteInterceptor {
+
+    public SemanticQueryRewriteInterceptor() {}
+
+    @Override
+    public QueryBuilder interceptAndRewrite(QueryRewriteContext context, QueryBuilder queryBuilder) {
+        QueryBuilder rewritten = queryBuilder;
+        String fieldName = getFieldName(queryBuilder);
+        ResolvedIndices resolvedIndices = context.getResolvedIndices();
+        if (resolvedIndices != null) {
+            InferenceIndexInformationForField indexInformation = resolveIndicesForField(fieldName, resolvedIndices);
+            if (indexInformation.getInferenceIndices().isEmpty()) {
+                // No inference fields were identified, so return the original query.
+                return rewritten;
+            } else if (indexInformation.nonInferenceIndices().isEmpty() == false) {
+                // Combined case where the field name requested by this query contains both
+                // semantic_text and non-inference fields, so we have to combine queries per index
+                // containing each field type.
+                rewritten = buildCombinedInferenceAndNonInferenceQuery(queryBuilder, indexInformation);
+            } else {
+                // The only fields we've identified are inference fields (e.g. semantic_text),
+                // so rewrite the entire query to work on a semantic_text field.
+                rewritten = buildInferenceQuery(queryBuilder, indexInformation);
+            }
+        }
```

Comment:
```suggestion
        if (resolvedIndices == null || indexInformation.getInferenceIndices().isEmpty()) {
            // No inference fields were identified, so return the original query.
            return rewritten;
        }
        if (indexInformation.nonInferenceIndices().isEmpty() == false) {
            // Combined case where the field name requested by this query contains both
            // semantic_text and non-inference fields, so we have to combine queries per index
            // containing each field type.
                return buildCombinedInferenceAndNonInferenceQuery(queryBuilder, indexInformation);
        }
        // The only fields we've identified are inference fields (e.g. semantic_text),
        // so rewrite the entire query to work on a semantic_text field.
        return buildInferenceQuery(queryBuilder, indexInformation);
```

Not a big deal, but to me, if its possible, early exit returning is easier to follow.

But no need to change, its just a taste thing.